Assorted touch fixes

yol opened the pull request:
I discovered issues with the touch subsystem while doing Wayland implementation tests (pkerling/xbmc#43), some of which are fixed by this PR

## Description
* Inertial scroll
  * Math does not add up (70b930fa) resulting in abrupt stop
  * Scrolling is not actually "inertial", i.e. if user starts touch, scrolls around a bit, then holds the finger for one second and releases, inertial scrolling will still kick in despite there not being any movement - very annoying (1d1c69f)
* Gesture detection
  * When windowing does not provide size, detection is way too sensitive and often turns taps into pan gestures (843e2b41)
  * Multi-pointer taps miss the check for whether the pointer moved even when size is provided (3d83da4) and so also frequently get turned into gestures
* Gestures are not aborted on window switch, resulting in unusable settings windows (because list control does not release exclusive mouse mode) when two-finger swipe gesture is used to go back to the previous window (f51f7bbd)

Rest of the commits are cleanup and C++11.

## Motivation and Context
Improve touch experience

## How Has This Been Tested?
Run-tested on Linux/Wayland

Android testing would be much appreciated. I can do it if someone tells me how I can get the resulting apk from jenkins?

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

pkerling committed:
Remove unused touch function

Signed-off-by: Philipp Kerling <pkerling@casix.org>

pkerling committed:
Remove unneeded memset usage

With C++11, zero-initialization is guaranteed.

Signed-off-by: Philipp Kerling <pkerling@casix.org>

pkerling committed:
Replace old-style casts in touch action handler

Signed-off-by: Philipp Kerling <pkerling@casix.org>

pkerling committed:
Replace plain pointers with RAII

Signed-off-by: Philipp Kerling <pkerling@casix.org>

pkerling committed:
Replace C arrays with std::array

pkerling committed:
Replace macro with constant

pkerling committed:
Use C++ math

pkerling committed:
Move touch size determination to GenericTouchInputHandler

Now all platforms can benefit from it equally. Otherwise some platforms
might send size 0.0 (default value) and end up with weird behavior
because touches get classified as moving even if they move only in
the sub-millimeter range.

Signed-off-by: Philipp Kerling <pkerling@casix.org>

pkerling committed:
Fix missing check for touch movement condition for multi-touch

Otherwise multi-finger taps are immediately classified as gesture
when one finger moves very slightly.

Signed-off-by: Philipp Kerling <pkerling@casix.org>

pkerling committed:
Send and handle gesture abort events

Contrary to GESTURE_END, GESTURE_ABORT does not carry any information
and just informs the controls that a running gesture has been aborted
e.g. by the windowing system.

Signed-off-by: Philipp Kerling <pkerling@casix.org>

pkerling committed:
Fix handling of touch gestures that are in progress when switching windows

If windows are switched while a touch gesture is active, the old window
would not get the GESTURE_END event and thus e.g. not release its
exclusive mouse control state, while the new window gets gesture events
for a gesture it does not even know has started.

So when switching windows, inform the old window that the touch gesture
has been aborted, and do not relay any gesture events to the new
window until a new gesture starts.

Signed-off-by: Philipp Kerling <pkerling@casix.org>

pkerling committed:
Inertial scrolling: Remove jerk

Inertial scrolling code included the decrease of the deacceleration
during the scrolling, resulting in non-zero jerk - but the math used
in the calculations does not add up. Also the conditions for checking
the time during which to apply the decrease are mixed up.
So go back to plain deacceleration with zero jerk, it at least works
reliably. If this is still needed, it must be done differently with
correct math.

Signed-off-by: Philipp Kerling <pkerling@casix.org>

pkerling committed:
Provide additional information about touch events in CAction

Third pair of X/Y data for touch events was just being dropped in
GenericTouchActionHandler, but controls might need it.

Signed-off-by: Philipp Kerling <pkerling@casix.org>

pkerling committed:
Make inertial scrolling actually take inertia into account

To make sure that we only start inertial scrolling when warranted,
only the movement just before removing the finger from the surface
must be considered. The code now remembers the pans happening in the
last 200 ms and calculates average velocity accordingly.

Signed-off-by: Philipp Kerling <pkerling@casix.org>

MartijnKaijser commented:
http://jenkins.kodi.tv/job/BuildMulti-All/821/  builds will end up in test-builds

notspiff commented on xbmc/input/touch/generic/GenericTouchActionHandler.cpp:
```diff
@@ -179,8 +182,8 @@ void CGenericTouchActionHandler::focusControl(float x, float y)
 
-  newEvent.focus.x = (uint16_t)x;
-  newEvent.focus.y = (uint16_t)y;
+  newEvent.focus.x = static_cast<int> (std::round(x));
```
This is not equivalent.

pkerling replied:
Yes. I think that it should have been rounded in the first place.

Montellese commented on xbmc/input/touch/generic/GenericTouchInputHandler.cpp:
```diff
@@ -28,7 +28,13 @@
-#define TOUCH_HOLD_TIMEOUT  1000
+namespace
+{
+
```
Please remove the additional blank lines.

pkerling replied:
done

Montellese reviewed (approved):
Looks good apart from the minor. I can't comment on the inertial scrolling, that's @Memphiz' baby.

Memphiz commented:
Tested on ios - seems to work fine still. I am fine with removing the decrease of deacceleration - feels good enough without imo. I added it in the first place to mimic what apple does as good as possible (the 0.9 decrease factor was determined by stomach feeling ...

Paxxi commented on xbmc/input/linux/LinuxInputDevices.h:
```diff
@@ -87,9 +86,9 @@ class CLinuxInputDevice
-  int m_mt_x[TOUCH_MAX_POINTERS];
-  int m_mt_y[TOUCH_MAX_POINTERS];
-  TouchInput m_mt_event[TOUCH_MAX_POINTERS];
+  int m_mt_x[CGenericTouchInputHandler::MAX_POINTERS];
```
If we're modernizing it would make sense to make these `std::aray` but might be out of scope for this PR.

pkerling replied:
I have to stop somewhere or this never ends ;-)

Paxxi commented on xbmc/input/touch/generic/GenericTouchActionHandler.cpp:
```diff
@@ -161,43 +166,28 @@ int CGenericTouchActionHandler::QuerySupportedGestures(float x, float y)
-  newEvent.type = XBMC_SETFOCUS;
-  newEvent.focus.x = (uint16_t)x;
-  newEvent.focus.y = (uint16_t)y;
+  newEvent.focus.x = static_cast<int> (std::round(x));
```
why is this changed from uint16_t to int?

pkerling replied:
XBMC_SetFocusEvent has `int`s, why cut it to 16 bit? https://github.com/xbmc/xbmc/blob/master/xbmc/windowing/XBMC_events.h#L106

Paxxi commented on xbmc/input/touch/generic/GenericTouchInputHandler.cpp:
```diff
@@ -361,44 +363,40 @@ void CGenericTouchInputHandler::triggerDetectors(TouchInput event, int32_t point
 
-  std::set<IGenericTouchGestureDetector*> finishedDetectors;
-  for (std::set<IGenericTouchGestureDetector*>::const_iterator detector = m_detectors.begin(); detector != m_detectors.end(); ++detector)
+  for (auto it = m_detectors.begin(); it != m_detectors.end(); )
```
might as well switch this to `std::remove_if` when changing it.

pkerling replied:
`remove_if` does not work with `std::set` because it is ordered and `remove_if` needs to reorder elements. I see no strong reason why `m_detectors` would need to be a `std::set` though. Not like you can try to push two identical `unique_ptr`s anyway. Should we make it a `std::vector`?

Paxxi reviewed (commented):
I trust @Montellese review on the functional changes, added a few nitpicks that are optional.

Can try and test the changes on Windows over the weekend.

pkerling committed:
Activate C++11 math on Android/ARM32

std::round is not supported without the _GLIBCXX_USE_C99_MATH_TR1
flag that is included in the 64-bit but not the 32-bit NDK build.
To build on 32-bit ARM, always activate it.

MartijnKaijser commented:
It feels like it improves touch quite a lot. Hitting buttons is sometimes still hit and miss but haven't tested extensively yet. Selecting in listings is still quite difficult to get the correct item selected but that could be a skinning engine issue. The highlighted item while scrolling also seems a bit flaky.

fritsch commented:
One thing at a time, please. If this one brings us forward, what it does,
squash together and we are good to go.

With the signed integer take care no one uses the center in the middle :-)

Am 05.08.2017 8:50 vorm. schrieb "Martijn Kaijser" <notifications@github.com
>:

It feels like it improves touch quite a lot. Hitting buttons is sometimes
still hit and miss but haven't tested extensively yet. Selecting in
listings is still quite difficult to get the correct item selected but that
could be a skinning engine issue. The highlighted item while scrolling also
seems a bit flaky.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/xbmc/xbmc/pull/12617#issuecomment-320419887>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ABCfHdU9sNnb6L-jyIFZCwgNADetBUhdks5sVBC1gaJpZM4OsClm>
.

MartijnKaijser commented:
@fritsch i was pointing out that one of the issues is still kinda there. So that is still that "one thing at a time".

pkerling commented:
I've identified potential problems with the gesture abort on window switch:

* I handled the gesture events in `CGUIWindowManager::OnApplicationMessage` - but the inertial scrolling handler might modify the events afterwards (drop some), so I think handling this in `CGUIWindowManager::OnAction` instead would be better. I'll add a fixup.
* What about other events that cause the focus and exclusive mouse control of a control to be broken no matter what it says - like dialogs (I get unresponsive dialogs from time to time, but haven't been able to pinpoint the problem)? Should we send gesture abort and inhibit likewise when `RegisterDialog` is called? Always or only under some circumstances, like only for modal dialogs or only when the topmost dialog changes?
  * Speaking of order: Some functions in window manager sort the dialogs based on their render order, like `RenderPass`. But most do not. `OnAction` for example doesn't do it. Won't that cause a mismatch between the topmost dialog on screen and the one that receives input?

pkerling committed:
Fix invalidating old pointer on multi-touch up

Without resetting the pointer, subsequent single-touch actions could
be misinterpreted as gestures due to the "moving" flag of the
old pointer still being set and valid.

MartijnKaijser commented:
@Memphiz could you try again on iOS if all still works? Android has improved a lot with this. @Paxxi if you could try on Windows?

Memphiz commented:
Is there a test build alteady?

MartijnKaijser commented:
Building now

pkerling commented:
squashed

Memphiz commented:
Behaves fine on iOS imo.

pkerling commented:
forgot to reorder the ARM32 fixed to before the std::round usage, so once again...
